android/app/src/main/java/io/peggyjo/peggymeter/PDKListener.java

```diff
@@ -2,17 +2,85 @@
 
 import android.content.Context;
 import android.os.Bundle;
+import android.support.annotation.NonNull;
 import android.util.Log;
-
 import com.audacious_software.passive_data_kit.generators.Generators;
 
-public class PDKListener implements Generators.GeneratorUpdatedListener {
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.firestore.DocumentReference;
+import com.google.firebase.firestore.FirebaseFirestore;
+import com.google.firebase.firestore.WriteBatch;
+
+
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.Date;
+
+
+public class PDKListener implements Generators.GeneratorUpdatedListener,
+        OnCompleteListener<Void> {
+    private HashMap<String, Long> timestamps = new HashMap<>();
+    private  final FirebaseFirestore db = FirebaseFirestore.getInstance();
+
     public void onGeneratorUpdated(String identifier, long timestamp, Bundle data) {
-        Log.i("PDK listener", "id: " + identifier + " | ts: " + timestamp + " | data.size = " + data.size());
-        // TODO: implement uploading to Firebase.
+       //Log.i("PDK listener", "id: " + identifier + " | ts: " + timestamp + " | data.size = " + data.size());
+        if (!timestamps.containsKey(identifier)) {
+            timestamps.put(identifier, timestamp);
+        } else {
+            Date lastEntry = new Date(timestamps.get(identifier));
+            Date current = Calendar.getInstance().getTime();
+
+            long deltaMins = calculateDeltaMins(current, lastEntry);
+            Log.i(identifier, "" + deltaMins);
+
+            if (deltaMins > 20) {
+                // TODO: implement uploading to Firebase.
+
+                //Sample generator data
+               String uid = FirebaseAuth.getInstance().getCurrentUser().getUid();
```

Comment:
Looks like something's wrong with the indentation here: this line and below.
Can you check please?

Reply:
Yeah.